flink-clients/src/main/java/org/apache/flink/client/LocalExecutor.java

```diff
@@ -172,29 +174,40 @@ public JobExecutionResult executePlan(Plan plan) throws Exception {
 
 				// start the cluster for us
 				start();
-			}
-			else {
+			} else {
 				// we use the existing session
 				shutDownAtEnd = false;
 			}
 
-			try {
-				Configuration configuration = this.flink.configuration();
+			Configuration configuration = this.flink.configuration();
 
-				Optimizer pc = new Optimizer(new DataStatistics(), configuration);
-				OptimizedPlan op = pc.compile(plan);
+			Optimizer pc = new Optimizer(new DataStatistics(), configuration);
+			OptimizedPlan op = pc.compile(plan);
 
-				JobGraphGenerator jgg = new JobGraphGenerator(configuration);
-				JobGraph jobGraph = jgg.compileJobGraph(op, plan.getJobId());
+			JobGraphGenerator jgg = new JobGraphGenerator(configuration);
+			JobGraph jobGraph = jgg.compileJobGraph(op, plan.getJobId());
 
-				boolean sysoutPrint = isPrintingStatusDuringExecution();
-				return flink.submitJobAndWait(jobGraph, sysoutPrint);
-			}
-			finally {
-				if (shutDownAtEnd) {
-					stop();
+			boolean sysoutPrint = isPrintingStatusDuringExecution();
+
+
+			JobListeningContext jobListeningContext = flink.submitJob(jobGraph, sysoutPrint);
+			JobClientEager jobClient = new JobClientEager(jobListeningContext);
+
+			Runnable cleanup = new Runnable() {
+				@Override
+				public void run() {
+					if (shutDownAtEnd) {
```

Comment:
We could but it wouldn't make any semantic difference since the enclosed variable must be final.